adr/0011-sap-system-database-aggregate-split.md

```diff
@@ -0,0 +1,33 @@
+
+In this scenario, when a multi-tenant database is discovered, it only can manage a unique SAP system and its application instances. Therefore, a multi-tenant scenario is not supported.
+
+This limitation must be overcome, as a multi-tenant scenario is pretty common, and many deployments have them.
+
+## Decision
+
+The currently existing `sap_system.ex` aggregate will be split into seperate modules, creating a new `database.ex` aggregate that will handle databases individually. This way, the lifecycle of a database is extracted from the lifecycle of the SAP system, making them independent. Having this, a new 1 to N relationship will be created between the database and SAP system aggregates. Now the read models are more precise. From now on, the SAP system ID will be calculated by combining the database ID and its tenant name, creating a new and unique ID for each SAP system.
+
+The relationship between database and SAP systems will continue existing in many senses, and most of the related events will be handled by event handlers. For example, when a database is deregistered, or its health updated, these events will be handled, and once they happen, they will affect the SAP system aggregate via commands. At the end, both aggregates continue being pure and connected through this kind of mechanism.
+
+All of this forces us to handle previously existing events and aggregate roll-up snapshots, so most of the events related to a database will be upcasted. Besides this, some additional exceptions might arise, as SAP system and database don't share their ID anymore, but they did in the past, so previously stored data (as the deregistration process manager) must be taken into consideration.
+
+In order to deal with all of these old events, besides upcasting, we have considered a couple of options when old events and IDs are mixed up. Initially, [prefixing](https://hexdocs.pm/commanded/Commanded.Commands.Router.html#module-define-aggregate-identity) the `sap_system.ex` aggregate and routing it with a `sap-system-` prefix looked a good option. It removes issues that could arise when SAP system commands with old IDs, now belonging to databases, are dispatched, as they are routed with the prefix together with the ID. We have discarded this option, as it converts the SAP system stream different from the others, and hides a bit what's going on in these scenarios. We decided against this solution, as it realistically only occours when the `DeregisterApplicationInstance` is dispatched by the deregistration process manager. Therefore, we will just handle this particular command. The command will be enriched in a middleware protocol, so commands with IDs, now belonging to databases, will be ignored.
```

Comment:
Did you mean different as an adjective? 

> as it converts the SAP system stream different from the others

This could sound nicer, what do you think ? 

> as it converts the SAP system stream **differently** from the others

Reply:
I meant that the SAP system stream would be different compared with the others